ramips: add support for Cudy X6

alessioprescenzo opened the pull request:
Specifications:
 - SoC: MediaTek MT7621
 - RAM: 256 MB
 - Flash: 32 MB
 - WiFi: MediaTek MT7915E
 - Switch: 1 WAN, 4 LAN (Gigabit)
 - Ports: 1 USB 3.0
 - Buttons: Reset, WPS
 - LEDs: Power, System, Wan, Lan 1-4, WiFi 2.4G, WiFi 5G, WPS, USB
 - Power: DC 12V 1A tip positive

Installation:

Download and flash the manufacturer's built OpenWRT image available at
http://www.cudytech.com/openwrt_software_download
Install the new OpenWRT image via luci (System -> Backup/Flash firmware)
Be sure to NOT keep settings. The force upgrade may need to be checked
due to differences in router naming conventions.

Recovery:
 - Loads only signed manufacture firmware due to bootloader RSA verification
 - serve tftp-recovery image as /recovery.bin on 192.168.1.88/24
 - connect to any lan ethernet port
 - power on the device while holding the reset button
 - wait at least 8 seconds before releasing reset button for image to
   download

Signed-off-by: Alessio Prescenzo <alessioprescenzo@gmail.com>

musashino205 commented on target/linux/ramips/dts/mt7621_cudy_x6.dts:
```diff
@@ -0,0 +1,166 @@
+};
+
+&gmac0 {
+	mtd-mac-address = <&bdinfo 0xde00>;
```
```mtd-mac-address``` property was replaced to NVMEM implementation.
see https://github.com/openwrt/openwrt/commit/06bb4a5018cd84dc4553e71c273e2daae2c996ab

musashino205 commented on target/linux/ramips/dts/mt7621_cudy_x6.dts:
```diff
@@ -0,0 +1,166 @@
+			status = "okay";
+			label = "wan";
+			mtd-mac-address = <&bdinfo 0xde00>;
+			mtd-mac-address-increment = <1>;
```
```mtd-mac-address-increment``` property was replaced to ```mac-address-increment```.
see https://github.com/openwrt/openwrt/commit/d284e6ef0f06551bad5696f16f26e6ff27b065d4

musashino205 commented on target/linux/ramips/image/mt7621.mk:
```diff
@@ -281,6 +281,16 @@ define Device/cudy_wr2100
 
+define Device/cudy_x6
+  $(Device/dsa-migration)
+  IMAGE_SIZE := 32448k
```
```32256k``` (```0x1F80000```)?

alessioprescenzo replied:
I've made all the changes required except this, i don't get what i should do, sorry :(

Before commit and push i want to fix everything

musashino205 replied:
Currently, ```IMAGE_SIZE``` is defined as ```32448k``` (= 0x1FB0000) but the size of "firmware" partition is defined as ```0x1f80000``` (= 32256k). ```IMAGE_SIZE``` basically needs to fit the size of "firmware" partition.

alessioprescenzo replied:
Got that, thanks for the explanation ur saving my day :D

alessioprescenzo commented:
I have a small problem. With the changed dts file now my LAN mac-address is random at every boot. Did i do smth wrong?

musashino205 commented:
Perhaps you need to use ```bdinfo``` instead of ```factory``` for MAC addresses.

https://github.com/openwrt/openwrt/pull/4582/commits/ece96167434ebece5e2006ac0f8282b805fdcc18#diff-2e939dd0265c59a983dfae2e08b6ab48a8ca0e37c772bd1cce2b88fef41d5af2L125-L129

alessioprescenzo commented:
For sure you are right, i've noticed and changed that partition name 10 minutes after my comment, tested and working, now with the correct mac :)

musashino205 commented on target/linux/ramips/dts/mt7621_cudy_x6.dts:
```diff
@@ -0,0 +1,178 @@
+};
+
+&bdinfo {
+ 	compatible = "nvmem-cells";
+ 	#address-cells = <1>;
+ 	#size-cells = <1>;
+
+ 	macaddr_bdinfo_de00: macaddr@de00 {
+ 		reg = <0xde00 0x6>;
```
Please delete a space at the beginning of these line.

alessioprescenzo replied:
on my end i don't see spaces at the beginning of those line, it's just a tab (/t)

musashino205 commented on target/linux/ramips/dts/mt7621_cudy_x6.dts:
```diff
@@ -0,0 +1,178 @@
+ 	macaddr_bdinfo_de00: macaddr@de00 {
+ 		reg = <0xde00 0x6>;
+	};
+};
```
Please add a line break at the end of file.

musashino205 commented:
You need to squash all commits to single commit.
see: https://openwrt.org/docs/guide-developer/working-with-github-pr#squashing_commits

alessioprescenzo committed:
ramips: add support for Cudy X6

Specifications:

SoC: MediaTek MT7621
RAM: 256 MB
Flash: 32 MB
WiFi: MediaTek MT7915E
Switch: 1 WAN, 4 LAN (Gigabit)
Ports: 1 USB 3.0
Buttons: Reset, WPS
LEDs: Power, System, Wan, Lan 1-4, WiFi 2.4G, WiFi 5G, WPS, USB
Power: DC 12V 1A tip positive
Installation:

Download and flash the manufacturer's built OpenWRT image available at
http://www.cudytech.com/openwrt_software_download
Install the new OpenWRT image via luci (System -> Backup/Flash firmware)
Be sure to NOT keep settings. The force upgrade may need to be checked
due to differences in router naming conventions.

Recovery:

Loads only signed manufacture firmware due to bootloader RSA verification
serve tftp-recovery image as /recovery.bin on 192.168.1.88/24
connect to any lan ethernet port
power on the device while holding the reset button
wait at least 8 seconds before releasing reset button for image to
download

Signed-off-by: Alessio Prescenzo <alessioprescenzo@gmail.com>

alessioprescenzo commented:
I've messed up a lot those commits and i seriously don't know how to revert this, can i close this this pull request and create another clean on? Sorry to bother

musashino205 commented:
I will explain from the current state in GitHub:

1. backup current development branch
  ```git branch Cudy-X6.back Cudy-X6```
  this creates a branch "Cudy-X6.back" from "Cudy-X6"

2. reset commits to before 19f1187689
  ```git reset --hard 19f1187689^```

3. add-back commit b6572e52a0
  ```git cherry-pick b6572e52a0```

4. add missing commit title
  ```git commit --amend```
  and insert title to the first line: ```ramips: add support for Cudy X6```

5. push changes to GitHub
  ```git push -f```

alessioprescenzo commented:
Thanks again for the help, cherry-picks returns a merge conflict: 
<img width="724" alt="image" src="https://user-images.githubusercontent.com/43753797/134382820-6e951562-885b-425b-83fc-8ad40664434b.png">

alessioprescenzo commented:
Thank you a lot, I've learned smth new thx to you today 👍  😄

musashino205 commented:
Oops, there is a mistake in the commit title, ```rampis```...

```
git commit --amend
# fix title and save/exit editor
git push -f
```

alessioprescenzo commented:
Everything should be fine now :D

adschm commented:
@musashino205 Thanks for helping me out here :-)

adschm commented on target/linux/ramips/dts/mt7621_cudy_x6.dts:
```diff
@@ -0,0 +1,178 @@
+
+&state_default {
+	gpio {
+		groups = "wdt", "i2c", "jtag";
```
Don't see any GPIOs for i2c?

alessioprescenzo replied:
I need a package to see them right? RN I'm cherry-picking my packages so there's a high probability that if this needs a package I don't have it installed (I'm saying that because I've seen the i2c section in menuconfig)

adschm replied:
No, essentially you just look which GPIOs you have used above and then put the corresponding groups here.
I've once compiled and Excel document for that.

You more or less have to extract the information from this file: https://github.com/torvalds/linux/blob/745ec436de7223fbe086be5123fe3014c4220611/drivers/pinctrl/ralink/pinctrl-mt7621.c or from the documentation on the board.

alessioprescenzo replied:
I'll look into it soon, meanwhile the manufacturer let you download this 
<img width="752" alt="image" src="https://user-images.githubusercontent.com/43753797/134775765-12ce0c6f-969c-481e-a043-ea040c103136.png">

idk if it's useful or not, just sharing my documentation.

adschm commented:
All MAC addresses correct (i.e. matching vendor firmware) now?

alessioprescenzo commented:
> All MAC addresses correct (i.e. matching vendor firmware) now?

Yes, now mac addresses are correct

s-2 commented:
> The force upgrade may need to be checked
> due to differences in router naming conventions.

Since this device seems to be based on a previous version of OpenWrt, could the OEM model name be added to image metadata via `SUPPORTED_DEVICES`?

rola25 commented:
@alessioprescenzo Do you have a precompiled firmware for Cudy X6 somewhere lying around? I have two X6 - I could test it and give you information if something is not working as it should ...

alessioprescenzo commented:
> @alessioprescenzo Do you have a precompiled firmware for Cudy X6 somewhere lying around? I have two X6 - I could test it and give you information if something is not working as it should ...

Hi, I do have a firmware that actually I regularly use on x6s (i also have two with one of them being  802.11r/k/v dumb AP) preinstalled with what I use (ip-full for VLANs, Adblock for well ad-blocking, and WireGuard VPN, and LuCi obv). I can share that if you want :)

rola25 commented:
@alessioprescenzo  Yes that would be fine. Can you share it please?

rola25 commented:
@alessioprescenzo I managed to compile your Cudy-X6 branch by myself, thanks anyway, it works

Nomorerd commented:
What's the process now? Is it official released?

adschm commented:
> > All MAC addresses correct (i.e. matching vendor firmware) now?
> 
> Yes, now mac addresses are correct

Please add a table like this to commit message: https://openwrt.org/docs/guide-developer/device-support-policies#mac_addresses

And this i2c in state_default still needs to be treated ...

brainchild0 commented:
Thanks to everyone who contributed.

Since this work seems close to   nearing completion (apparently pending a few housekeeping issues, and reviews), would anyone be able to post a clean build of the results of the efforts so far, for loading into my devices?

F5BJR reviewed (commented):
**Who could add the CUDY-6 (Wifi 6 AX) to snapshots?**

I have had two of these boxes for a while and thanks to our friend now the mac addresses of the 2.4 GHz and 5GHz wifi are consistent

**Test:**
WIFI 2.4 Ghz & 5Ghz run fine 
Lan port number match with the sticker & Wan +1
Wifi mac address are correct match with the sticker + 2 & 3 **thanks ur friend blocktrron developer**

dts file is correct  

The additions are as follows:

add in file **target/ramips/mt7621/base-files/etc/hotplug.d/ieee80211/10_fix_wifi_mac**
just after case "$board" in

```
cudy,x6)
		hw_mac_addr="$(mtd_get_mac_binary factory 0x4)"
		[ "$PHYNBR" = "0" ] && macaddr_add $hw_mac_addr "2" > /sys${DEVPATH}/macaddress
		[ "$PHYNBR" = "1" ] && macaddr_add $hw_mac_addr "3" > /sys${DEVPATH}/macaddress
		;; 
```

brainchild0 commented:
I am eager to begin operation on my device, and am looking for the most straightforward way to build.

I am new to the project, and wondered whether I might ask for a bit of assistance, and hoped posting my questions in my current location is not objectionable.

What I have done is [applied the revision of the current request over tag `v21.02.1`](https://github.com/brainchild0/openwrt/commits/Cudy-X6), the recent release. In this way I hoped to create an image representing a stable release,   the same as   would have been created  had support for the device been added before the release. Following the same idea, I hoped   to apply  all of the default package selections, and other options. 

I searched  for a straightforward and clear way to achieve the  effect I had described, hoping to bypass any interactive steps, where I might easily make mistakes,  such as invoking `menuconfig`.  I also searched for a way to utilize the new entry `#define Device/cudy_x6`, though was not much helped by the resources I found.    Based on one source, I thought  it might be correct to pass  `PROFILE=Device/cudy_x6` to the invocation of  `make`, but this idea did not lead to any  success. Instead, the build process broke near the end.

I was able to build without any options, or any invocation of `menuconfig`, which at least validates my build environment.


I did notice a warning, as follows, and wondered whether it expressed a problem, or rather was safe to ignore:

```
WARNING: your configuration is out of sync. Please run make menuconfig, oldconfig or defconfig!
```
What is the most straightforward way to build the target for the new device, Cudy x6, based on my steps so far?

LiftedStarfish commented:
What's the status on this? How long should I expect it to take for this to actually be merged?

LiftedStarfish commented:
Is this good now? Can I install it on my router?

F5BJR commented:
Yes no problem envoyé : 10 mai 2022 à 19:32de : LiftedStarfish ***@***.***>à : openwrt/openwrt ***@***.***>cc : F5BJR ***@***.***>, Comment ***@***.***>objet : Re: [openwrt/openwrt] ramips: add support for Cudy X6 (#4582) Is this good now? Can I install it on my router?—Reply to this email directly, view it on GitHub, or unsubscribe.You are receiving this because you commented.Message ID: ***@***.***>

waja commented:
It was included into master, but not into openwrt-22.03 branch. So it seems it will still be missing with the 22.03 release. :(

brainchild0 commented:
For someone running the current vendor image  (e.g.    OpenWrt 21.02-SNAPSHOT r15949-b2c9a8741f),   what is the best path for upgrade?

Will the   *sysupgrade* feature work correctly, even though the currently-installed image is not an official project build?   Will settings be preserved for consistent operation before and after upgrade?

tomck commented:
According to @julyworlds https://github.com/julyworlds/openwrt-cudy-x6-firmware, @brainchild0, yes just upgrading to the snapshot will work.  (I personally just upgraded to the stable version provided by their CI and it seems to work fine.)

For a link to the current snapshot bin, visit the wiki: https://openwrt.org/toh/hwdata/cudy/cudy_x6

(Note, there is no LuCi with the snapshots, you have to [install it via SSH](https://openwrt.org/docs/guide-user/luci/luci.essentials))

brainchild0 commented:
@tomck, So the settings will be lost, and I will need to install the web interface manually after the upgrade?

tomck commented:
> @tomck, So the settings will be lost, and I will need to install the web interface manually after the upgrade?

Yes, always assume the settings will be lost, according to julyworlds you need to discard settings when upgrading from manufacturer OpenWRT to real OpenWRT.

brainchild0 commented:
@tomck May settings be saved effectively by capturing them  in an external file that may be imported after booting from the replacement image?

tomck commented:
@brainchild0 Probably. I don't work with the OpenWRT team, was just sharing my experience, and can make no guarantees. If you are really worried, screenshot all your settings pages first so you have a reference.
